zerver/lib/retention.py

```diff
@@ -244,23 +243,28 @@ def run_archiving_in_chunks(message_id_chunks: Iterator[List[int]]) -> int:
             except StopIteration:
                 break
 
+            archive_transaction = ArchiveTransaction.objects.create(
+                type=ArchiveTransaction.RETENTION_POLICY_BASED, realm=realm
+            )
+            ArchivedMessage.objects.filter(id__in=chunk).update(archive_transaction=archive_transaction)
```

Comment:
Hmm, I guess this workflow makes it kinda awkward to pass the `archive_transaction` when generating `message_id_chunks`; given we're in a `transaction.atomic` I suppose the ordering isn't a correctness problem, but it would be faster if we did it another way.  

Which I think we could if we had `(archive_transaction, chunk) = next(message_id_chunks)` be the iterator API?

Reply:
Oh, it looks like you did a different refactor with that goal anyway later in the branch?

Reply:
Yes, this code gets refactored into a much better form I think, in " retention: Combine run_message_batch_query and run_archiving_in_chunks."